Lib/asyncio/__main__.py

```diff
@@ -0,0 +1,100 @@
+import ast
+import asyncio
+import code
+import concurrent.futures
+import inspect
+import sys
+import threading
+import types
+
+from . import futures
+
+
+class AsyncIOInteractiveConsole(code.InteractiveConsole):
+
+    def __init__(self, locals, loop):
+        super().__init__(locals)
+        self.compile.compiler.flags |= ast.PyCF_ALLOW_TOP_LEVEL_AWAIT
+
+        self.loop = loop
+
+    def runcode(self, code):
+        future = concurrent.futures.Future()
+
+        def callback():
+            global repl_future
+
+            func = types.FunctionType(code, self.locals)
+            try:
+                coro = func()
+            except BaseException as ex:
+                future.set_exception(ex)
+                return
+
+            if not inspect.iscoroutine(coro):
+                future.set_result(coro)
+                return
+
+            try:
+                repl_future = self.loop.create_task(coro)
+                futures._chain_future(repl_future, future)
```

Comment:
Isn't `asyncio.run_coroutine_threadsafe` exists for replacing almost the whole `callback()` code?

Reply:
Almost, but not all.  The idea here is that if the user is executing `something()` as opposed to `await otherthing()`, the former must execute in the main thread too.

Reply:
That's why I needed to copy/paste some of the `run_coroutine_threadsafe` implementation to make it work for our use case. Since it's all in asyncio, I see no problem in that.

Reply:
Ok, no problem.